### ISSUE TITLE
[no-unused-vars] False positive for generic parameters

### ISSUE THREAD

**Issue by edilson** [typescript-eslint/typescript-eslint#3328]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": ["error", {
      "argsIgnorePattern": "_"
    }]
  }
}
```

```TS
recover<T>(key: string): Promise<any>;
```

**Expected Result**

The generic type parameter not to be identified as an unused var.

**Actual Result**

The generic type parameter is being identified as an unused var.

**Additional Info**

Thats the error throwed:

```
8:11  error  'T' is defined but never used  @typescript-eslint/no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.14.1` |
| `@typescript-eslint/parser`        | `4.14.1` |
| `TypeScript`                       | `4.1.5` |
| `ESLint`                           | `7.18.0` |
| `node`                             | `14.4.0` |

**Comment by bradzacher** [comment]
From your example - it's not being used, so OFC it's being flagged.

**Comment by edilson** [comment]
That's weird because when I was using the 2.34.0 version it didn't give me the error but updating to 4.14.1 this error occurs.

Maybe I should've put that this function is declared at an interface like this:

```TS
export default interface ICacheProvider {
  recover<T>(key: string): Promise<any>;
}
```

**Comment by bradzacher** [comment]
Check out the release notes. We fixed the unused variable detection between those two versions.

Again - your generic ***is not used at all***.
If it's not being used - then it's an unused variable, hence it is being reported.
